### PR TITLE
News

### PR DESCRIPTION
**Pull request by modelrailroader**
#2791 and general improvements

### CODE REVIEW

File: phpmyfaq/admin/assets/src/api/news.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  } catch (error) {
+    console.error('Error posting news data:', error);
+    throw error;
+  }
+};
+
+export const deleteNews = async (csrfToken, id) => {
+  try {
+    const response = await fetch('api/news/delete', {
+      method: 'POST',
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/api/news.js:49]: It should be DELETE instead of POST

File: phpmyfaq/admin/assets/src/api/news.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  } catch (error) {
+    console.error('Error deleting news data:', error);
+    throw error;
+  }
+};
+
+export const updateNews = async (data = {}) => {
+  try {
+    const response = await fetch('api/news/update', {
+      method: 'POST',
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/api/news.js:80]: It should be PUT instead of POST

File: phpmyfaq/admin/assets/src/content/news.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import { TinyMCE } from 'tinymce';
+import tinymce from 'tinymce/tinymce';
+
+export const handleAddNews = () => {
+  const submit = document.getElementById('submitAddNews');
+  if (submit) {
+    submit.addEventListener('click', async (event) => {
+      event.preventDefault();
+      let target = '';
+      document.querySelectorAll('#target').forEach(function (item) {
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/content/news.js:28]: You can use an arrow function here:

`document.querySelectorAll('#target').forEach(item) => {`

File: phpmyfaq/admin/assets/src/content/news.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      }
+      await addNews(data);
+    });
+  }
+}
+
+export const handleNews = () => {
+  const deleteNewsButton = document.getElementById('deleteNews');
+  if (deleteNewsButton) {
+    document.querySelectorAll('#deleteNews').forEach(function(item) {
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/content/news.js:57]: same here, please use an arrow function

File: phpmyfaq/admin/assets/src/content/news.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      item.addEventListener('click', function(event) {
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/content/news.js:58]: same here, please use an arrow function

File: phpmyfaq/admin/assets/src/content/news.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        modal.show();
+      });
+    });
+    document.getElementById('pmf-delete-news-action').addEventListener('click', async (event) => {
+      event.preventDefault();
+      const csrfToken = document.getElementById('pmf-csrf-token-delete').value;
+      const id = document.getElementById('newsId').value;
+      await deleteNews(csrfToken, id);
+    });
+    document.querySelectorAll('#activate').forEach(function (item) {
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/content/news.js:71]: same here, please use an arrow function

File: phpmyfaq/admin/assets/src/content/news.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  }
+}
+
+export const handleEditNews = () => {
+  const submit = document.getElementById('submitEditNews');
+  if (submit) {
+    submit.addEventListener('click', async (event) => {
+      event.preventDefault();
+      let target = '';
+      document.querySelectorAll('#target').forEach(function (item) {
```
Comment (thorsten) [inline phpmyfaq/admin/assets/src/content/news.js:85]: same here, please use an arrow function

File: phpmyfaq/admin/news.php [hunk 8]
```diff
@@ -38,565 +39,113 @@
...
+    'ad_news_delsuc' => Translation::get('ad_news_delsuc'),
+    'ad_news_updatesuc' => Translation::get('ad_news_updatesuc'),
+    'msgDeleteNews' => Translation::get('msgDeleteNews'),
+    'csrfToken_deleteNews' => Token::getInstance()->getTokenString('delete-news'),
+    'csrfToken_updateNews' => Token::getInstance()->getTokenString('update-news'),
+    'ad_entry_active' => Translation::get('ad_entry_active'),
+    'csrfToken_activateNews' => Token::getInstance()->getTokenString('activate-news')
+];
+
+$filterCreateIsoDate = new TwigFilter('createIsoDate', function ($string) {
```
Comment (thorsten) [inline phpmyfaq/admin/news.php:93]: We could move the 2 TwigerFilter classes to the Template folder and create new classes and files there.

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+class NewsController extends AbstractController
+{
+    #[Route('admin/api/news/add')]
+    public function addNews(Request $request)
+    {
+        $this->userHasPermission(PermissionType::NEWS_ADD);
+        $data = json_decode($request->getContent());
+
+        $faqConfig = Configuration::getConfigurationInstance();
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:40]: You don't need this, you can get the configuration instance from

`$this->configuration`

as it's defined and instantiated in the AbstractController

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+            ->setDateEnd(new DateTime($dateEnd))
+            ->setLink($link ?? '')
+            ->setLinkTitle($linkTitle ?? '')
+            ->setLinkTarget($target ?? '')
+            ->setCreated(new DateTime());
+
+        if ($news->create($newsMessage)) {
+            return $this->json(['success' => Translation::get('ad_news_updatesuc')], Response::HTTP_OK);
+        } else {
+            return $this->json(['error' => Translation::get('err_NotAuth')], Response::HTTP_UNAUTHORIZED);
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:79]: This error is not an auth error but an error during saving the news data.

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+
+    #[Route('admin/api/news/delete')]
+    public function deleteNews(Request $request)
+    {
+        $this->userHasPermission(PermissionType::NEWS_DELETE);
+        $data = json_decode($request->getContent());
+
+        $faqConfig = Configuration::getConfigurationInstance();
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:89]: You don't need this, you can get the configuration instance from

`$this->configuration`

as it's defined and instantiated in the AbstractController

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+
+    #[Route('admin/api/news/update')]
+    public function updateNews(Request $request)
+    {
+        $this->userHasPermission(PermissionType::NEWS_DELETE);
+        $data = json_decode($request->getContent());
+
+        $faqConfig = Configuration::getConfigurationInstance();
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:111]: You don't need this, you can get the configuration instance from

`$this->configuration`

as it's defined and instantiated in the AbstractController

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        if (!Token::getInstance()->verifyToken('delete-news', $data->csrfToken)) {
+            return $this->json(['error' => Translation::get('err_NotAuth')], Response::HTTP_UNAUTHORIZED);
+        }
+
+        $deleteId = Filter::filterVar($data->id, FILTER_VALIDATE_INT);
+
+        if ($news->delete((int)$deleteId)) {
+            return $this->json(['success' => Translation::get('ad_news_delsuc')], Response::HTTP_OK);
+        } else {
+            return $this->json(['error' => Translation::get('err_NotAuth')], Response::HTTP_UNAUTHORIZED);
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:101]: This error is not an auth error but an error during saving the news data.

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            ->setDateEnd(new DateTime($dateEnd))
+            ->setLink($link ?? '')
+            ->setLinkTitle($linkTitle ?? '')
+            ->setLinkTarget($target ?? '')
+            ->setCreated(new DateTime());
+
+        if ($news->update($newsMessage)) {
+            return $this->json(['success' => Translation::get('ad_news_updatesuc')], Response::HTTP_OK);
+        } else {
+            return $this->json(['error' => Translation::get('err_NotAuth')], Response::HTTP_UNAUTHORIZED);
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:152]: This error is not an auth error but an error during saving the news data.

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        $id = Filter::filterVar($data->id, FILTER_VALIDATE_INT);
+        $status = Filter::filterVar($data->status, FILTER_SANITIZE_SPECIAL_CHARS);
+
+        if ($news->activate($id, $status)) {
+            return $this->json(['success' => Translation::get('ad_news_updatesuc')], Response::HTTP_OK);
+        } else {
+            return $this->json(['error' => Translation::get('err_NotAuth')], Response::HTTP_UNAUTHORIZED);
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:175]: This error is not an auth error but an error during saving the news data.

File: phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+
+    #[Route('admin/api/news/activate')]
+    public function activateNews(Request $request)
+    {
+        $this->userHasPermission(PermissionType::NEWS_EDIT);
+        $data = json_decode($request->getContent());
+
+        $faqConfig = Configuration::getConfigurationInstance();
```
Comment (thorsten) [inline phpmyfaq/src/phpMyFAQ/Controller/Administration/NewsController.php:162]: You don't need this, you can get the configuration instance from

`$this->configuration
`
as it's defined and instantiated in the AbstractController

**Review by thorsten — CHANGES_REQUESTED** [review]

File: phpmyfaq/admin/news.php (hunk 8 above)
Comment (modelrailroader) [reply phpmyfaq/admin/news.php:93]: I'll do it after merging together with refactoring the user.php and change it in all files.

**Comment by modelrailroader** [comment]
Fixed all required changes, explanation: see Discord.

**Pull request merged**